lua/buffertabs/init.lua

```diff
@@ -168,16 +172,48 @@ local function display_buffers()
     local max = U.get_max_width(data, cfg)
     width = U.get_position_horizontal(cfg, max, #data)
 
-    for idx, v in pairs(data) do
-        if #data == 1 and not cfg.show_single_buffer then
-            return
+    local buffer_count = #data
+
+    if cfg.show_single_buffer == false and buffer_count <= 1 then
+        return
+    end
+
+    if cfg.max_buffers > 0 and buffer_count > cfg.max_buffers then
+        return
+    end
+
+    -- It only makes sense to show the surrounding buffers if there are enough buffers to show
+    local minimum_buffers_to_show = 2 * cfg.surround_active_buffer + 1
+
+    if cfg.surround_active_buffer > 0 and buffer_count >= minimum_buffers_to_show then
+
+        -- Find the index of the active buffer
+        local active_index = nil
+        for idx, v in pairs(data) do
+            if v.active then
+                active_index = idx
+                break
+            end
         end
 
-        create_win(v.name, v.active, v.modified, idx)
+        local lowest_idx = active_index - cfg.surround_active_buffer
+        local highest_idx = active_index + cfg.surround_active_buffer
+
+        local total_shown = 0
+        for idx = lowest_idx, highest_idx do
+            total = total + 1
+            local buffer_idx = idx % buffer_count -- Wrap around to start of list
+            if buffer_idx <= 0 then buffer_idx = buffer_count + buffer_idx end -- Wrap around to end of list
+            local buffer_data = data[buffer_idx]
+            create_win(buffer_data.name, buffer_data.active, buffer_data.modified, total_shown)
+        end
```

Comment:
This is pretty messy.

Essentially we are just finding the neighbours either side of the active buffer (which sometimes involves wrapping around either side of the list).

We also keep track of the total amount of shown tabs so `create_win` knows where to put them